### PR TITLE
refactor: DSearch implementation 

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description
https://mask.atlassian.net/browse/MF-2691
https://mask.atlassian.net/browse/MF-2815
https://mask.atlassian.net/browse/MF-2830

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [x] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Comment by zhouhanseng** [comment]
My VPN is too unstable to do `pnpm i` to update the pnpm yaml. I will try it tomorrow.

File: packages/web3-state/src/NameService.ts [hunk 1]
```diff
@@ -56,7 +56,7 @@ export class NameServiceState<
         })
     }
 
-    async lookup(chainId: ChainId, name: string) {
+    lookup = async (chainId: ChainId, name: string) => {
```
Comment (guanbinrui) [inline packages/web3-state/src/NameService.ts]: What's the difference from the previous version?

Comment (hanszhoou) [reply packages/web3-state/src/NameService.ts]: The scope of `this`.

```
const lookup = NameService.lookup

lookup() // incorrect
```

File: packages/web3-providers/src/d-search/index.ts [hunk 2]
```diff
@@ -1,19 +1,93 @@
...
-    implements DSearchBaseAPI.Provider<ChainId, SchemaType>
-{
-    search(keyword: string, sourceType?: SearchSourceType): Promise<SearchResult<ChainId, SchemaType>> {
-        return Promise.resolve({
+const CHAIN_ID_LIST = [ChainIdEVM.Mainnet, ChainIdEVM.BSC, ChainIdEVM.Matic]
+
+export class DSearchAPI<ChainId = Web3Helper.ChainIdAll> implements DSearchBaseAPI.Provider<ChainId, NetworkPluginID> {
+    async search(
+        keyword: string,
+        options: {
```
Comment (guanbinrui) [inline packages/web3-providers/src/d-search/index.ts]: Remove `options` and import `@masknet/web3-shared-*` stead.

File: packages/web3-shared/base/src/specs/index.ts [hunk 3]
```diff
@@ -94,28 +94,23 @@ export enum SourceType {
...
     Domain = 'Domain',
     // e.g., $MASK #MASK
-    FungibleToken = 'FungibleToken',
-    // e.g., $PUNK #PUNK
-    NonFungibleToken = 'NonFungibleToken',
-    // e.g., $PUNK #PUNK
-    NonFungibleCollection = 'NonFungibleCollection',
+    TrendingTokenByKeyword = 'TrendingTokenByKeyword',
+    // e.g., 0x69af81e73a73b40adf4f3d4223cd9b1ece623074 (Mask Network)
+    TrendingTokenByAddress = 'TrendingTokenByAddress',
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: They are the same thing `FungibleToken`. And please don't remove `NonFungibleToken`. It's used in the next feature.

File: packages/web3-shared/base/src/specs/index.ts [hunk 4]
```diff
@@ -604,17 +609,31 @@ export interface NonFungibleTokenSecurity {}
...
+}
+export interface EOAResult<ChainId> extends Result<ChainId> {
+    type: SearchResultType.EOA
+    domain?: string
+    address: string
 }
 
-export interface AddressResult<ChainId> extends Result<ChainId> {
-    type: SearchResultType.Address
+export interface CAResult<ChainId> extends Result<ChainId> {
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: This `CAResult` duplicate with `FungibleTokenResult` or `NonFungibleTokenResult`.

Comment (guanbinrui) [reply packages/web3-shared/base/src/specs/index.ts]: The `DSearchAPI` should take responsibility for returning what a contract address belongs to. A rough `CAResult` is meanless for UI.

File: packages/web3-providers/src/types/Trending.ts [hunk 5]
```diff
@@ -187,7 +187,7 @@ export namespace TrendingAPI {
         /** Get all related coins with a keyword. */
         getCoinsByKeyword: (chainId: ChainId, keyword: string) => Promise<Coin[]>
         /** Get coin info by address. */
-        getCoinInfoByAddress: (chainId: ChainId, address: string) => Promise<CoinInfo | undefined>
+        getCoinInfoByAddress: (chainId: ChainId | undefined, address: string) => Promise<CoinInfo | undefined>
```
Comment (guanbinrui) [inline packages/web3-providers/src/types/Trending.ts]: Why accept an `undefined`?

File: packages/web3-providers/tests/DSearch/index.ts [hunk 6]
```diff
@@ -1,12 +1,11 @@
-import { FungibleTokenResult, NonFungibleCollectionResult } from '@masknet/web3-shared-base'
 import { describe, expect, test } from 'vitest'
 import { DSearchAPI } from '../../src/DSearch/index.js'
 
 /* cspell:disable */
 describe('DSearch test', () => {
     test('should return by sysmbol from specific list', async () => {
         const DSearch = new DSearchAPI()
-        const result = await DSearch.search('token:eth')
+        const result = await DSearch.searchToken('token:eth')
```
Comment (guanbinrui) [inline packages/web3-providers/tests/DSearch/index.ts]: Why? The `token:` prefix already indicates it is searching for a token result.

File: packages/web3-providers/src/NFTScan/apis/TrendingAPI.ts [hunk 7]
```diff
@@ -111,23 +110,26 @@ export class NFTScanTrendingAPI implements TrendingAPI.Provider<ChainId> {
...
-                    if (!collection?.contract_address) return undefined
-                    return { collection, chainId }
-                } catch {
-                    return undefined
-                }
-            }),
-            undefined,
-        )
-        if (!result) {
+        console.log({ chainId_ })
```
Comment (guanbinrui) [inline packages/web3-providers/src/NFTScan/apis/TrendingAPI.ts]:
```suggestion
```

File: packages/web3-providers/src/DSearch/index.ts [hunk 8]
```diff
@@ -90,7 +118,7 @@ export class DSearchAPI<ChainId = Web3Helper.ChainIdAll, SchemaType = Web3Helper
      * "address:0x"
      *
      */
-    async search(keyword: string): Promise<Array<SearchResult<ChainId, SchemaType>>> {
+    async searchToken(keyword: string): Promise<Array<SearchResult<ChainId, SchemaType>>> {
```
Comment (guanbinrui) [inline packages/web3-providers/src/DSearch/index.ts]: We search for more than tokens. The new name looks not correct to me.

File: packages/plugins/RSS3/src/SNSAdaptor/index.tsx [hunk 9]
```diff
@@ -34,24 +36,74 @@ const createActivitiesTabConfig = (label: string, props: FeedPageProps, priority
     }
 }
 
-const ActivitiesTabConfig: Plugin.SNSAdaptor.ProfileTab = createActivitiesTabConfig('Activities', {})
-const ActivitiesTabConfigInProfileCard: Plugin.SNSAdaptor.ProfileTab = createActivitiesTabConfig('Activities', {}, 2)
+const createSearchTabConfig = (
```
Comment (guanbinrui) [inline packages/plugins/RSS3/src/SNSAdaptor/index.tsx:39]: CC @UncleBill, Please take a look.

File: packages/web3-shared/base/src/specs/index.ts [hunk 10]
```diff
@@ -95,28 +95,33 @@ export enum SourceType {
     R2D2 = 'R2D2',
 }
 
-export enum SearchSourceType {
-    Main = 'Main',
-    Sidebar = 'Sidebar',
-}
-
-/** @deprecated use SearchSourceType stead */
+/** @deprecated use SearchResultType stead */
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Can we remove it now?

File: packages/mask/src/extension/popups/pages/Wallet/ConnectWallet/index.tsx [hunk 11]
```diff
@@ -78,7 +78,7 @@ const ConnectWalletPage = memo(() => {
             provider: Web3Helper.Web3ProviderDescriptor<NetworkPluginID.PLUGIN_EVM>,
         ) => {
             if (provider.type === ProviderType.MaskWallet) {
-                const connection = await Connection?.getConnection?.()
+                const connection = Connection?.getConnection?.()
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/ConnectWallet/index.tsx]: Missed one. I will do it in another PR.

File: packages/plugins/EVM/src/state/Connection/connection.ts [hunk 12]
```diff
@@ -154,10 +154,10 @@ class Connection implements EVM_Connection {
         }
     }
 
-    private getOptions(
+    private getOptions = (
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Connection/connection.ts]: It's poor maintainability. How can others know whether if need to use the arrow function? Why not use `bind`?

File: packages/web3-shared/base/src/specs/index.ts (hunk 10 above)
Comment (hanszhoou) [reply packages/web3-shared/base/src/specs/index.ts]: yep

File: packages/web3-providers/src/DSearch/index.ts (hunk 8 above)
Comment (hanszhoou) [reply packages/web3-providers/src/DSearch/index.ts]: `searchToken()` invoked by `search()`

File: packages/web3-shared/base/src/specs/index.ts [hunk 13]
```diff
@@ -95,28 +95,27 @@ export enum SourceType {
...
-    Domain = 'Domain',
-}
-
 export enum SearchResultType {
-    // e.g., 0xd8da6bf26964af9d7eed9e03e53415d37aa96045
-    Address = 'Address',
+    // e.g., 0x6a7122B831c2B79c508A978f73f2ee23171279B3
+    EOA = 'EOA',
+    // Contract Address, not sure is fungible or nonFungible token
+    CA='CA',
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Remove it.

File: packages/web3-shared/base/src/specs/index.ts [hunk 14]
```diff
@@ -95,28 +95,25 @@ export enum SourceType {
...
-    // e.g., $PUNK #PUNK
-    NonFungibleCollection = 'NonFungibleCollection',
+    // e.g., #APE
+    NonFungibleToken='NonFungibleToken',
+    NonFungibleCollection='NonFungibleCollection',
+    // hits none of all above.
+    Unknown = 'Unknown'
+}
+
+export enum SearchFungibleTokenResultSubType {
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Remove it.

File: packages/web3-providers/src/NFTScan/apis/TrendingAPI.ts [hunk 15]
```diff
@@ -111,23 +110,25 @@ export class NFTScanTrendingAPI implements TrendingAPI.Provider<ChainId> {
...
-                } catch {
-                    return undefined
-                }
-            }),
-            undefined,
-        )
-        if (!result) {
+        const result = chainId_
+            ? { collection: await this.getCollection(chainId_, id), chainId: chainId_ }
+            : await attemptUntil(
```
Comment (guanbinrui) [inline packages/web3-providers/src/NFTScan/apis/TrendingAPI.ts]: This API should serve to query one chain id each time.

File: packages/web3-providers/src/DSearch/index.ts [hunk 16]
```diff
@@ -90,11 +118,37 @@ export class DSearchAPI<ChainId = Web3Helper.ChainIdAll, SchemaType = Web3Helper
...
```
Comment (guanbinrui) [inline packages/web3-providers/src/DSearch/index.ts]:
```suggestion
    private async searchToken(keyword: string): Promise<Array<SearchResult<ChainId, SchemaType>>> {
```

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**